packages/typespec-vscode/src/vscode-cmd/openapi3-preview.ts

```diff
@@ -0,0 +1,323 @@
+import { readdir, rm } from "fs/promises";
+import * as semver from "semver";
+import * as vscode from "vscode";
+import logger from "../log/logger.js";
+import { getBaseFileName, getDirectoryPath, joinPaths } from "../path-utils.js";
+import { TspLanguageClient } from "../tsp-language-client.js";
+import { createTempDir, parseJsonFromFile, throttle } from "../utils.js";
+
+const TITLE = "Preview in OpenAPI3";
+
+export async function showOpenApi3(
+  uri: vscode.Uri,
+  context: vscode.ExtensionContext,
+  client: TspLanguageClient,
+) {
+  const selectedFile = uri?.fsPath ?? vscode.window.activeTextEditor?.document.uri.fsPath;
+  if (!selectedFile || !selectedFile.endsWith(".tsp")) {
+    logger.error(
+      "Please select a Typespec file",
+      selectedFile ? [`Currently '${selectedFile}' is selected`] : [],
+      {
+        showOutput: true,
+        showPopup: true,
+      },
+    );
+    return;
+  }
+
+  const mainTspFile = selectedFile.endsWith("main.tsp") ? selectedFile : await getMainTspFile();
+  if (mainTspFile === undefined) {
+    logger.error(`No 'main.tsp' file can be determined from '${selectedFile}' in workspace.`, [], {
+      showOutput: true,
+      showPopup: true,
+    });
+    return;
+  }
+
+  await loadOpenApi3PreviewPanel(mainTspFile, context, client);
+}
+
+async function getMainTspFile(): Promise<string | undefined> {
+  const files = await vscode.workspace.findFiles("**/main.tsp").then((uris) => {
```

Comment:
TraverseMainTspFileInWorkspace() in typespec-util?

Reply:
OK. I didn't realized that `"**/node_moduels/**"` is an exclusive pattern when I first read it.
Yes. We can use that.